flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/internal/ProducerPoolImpl.java

```diff
@@ -131,26 +131,33 @@ public void recycleByTransactionId(String transactionalId) {
         if (!earlierTransactions.isEmpty()) {
             for (String id : earlierTransactions.values()) {
                 ProducerEntry entry = producerByTransactionalId.remove(id);
-                recycleProducer(entry.getProducer());
+                recycleProducer(entry.getProducer(), false);
             }
             earlierTransactions.clear();
         }
     }
 
     @Override
     public void recycle(FlinkKafkaInternalProducer<byte[], byte[]> producer) {
-        recycleProducer(producer);
+        recycleProducer(producer, true);
         ProducerEntry producerEntry =
                 producerByTransactionalId.remove(producer.getTransactionalId());
         transactionalIdsByCheckpoint.remove(producerEntry.getCheckpointedTransaction());
     }
 
-    private void recycleProducer(@Nullable FlinkKafkaInternalProducer<byte[], byte[]> producer) {
+    private void recycleProducer(
```

Comment:
Nit: Maybe it makes sense to handle the success criteria `recycleByTransactionId` and do not call `recyleProducer` at all if the transaction wasn't finished successfully.

It removes the additional changes that are only caused by the signature change.

Reply:
It's a good idea but it makes the call-site messier. There are a total of 4 cases now:
* Producer is recycled in writer (abort loop). That would go well with your proposal.
* Committer regularly commits and recycles. That would also go well with your proposal.
* Special case 1: on recovery, committer commits and recycles the id, in that case the producer entry doesn't contain an internal producer and is null. So we would need to handle the null on call-site.
* Special case 2: on non-chained committer, we may end up with having more transactions in the state than there are running (see the comment in `recycleByTransactionId`). Here, we also may have null values. 

So in cases where the producer is null or reuse is false, it gets messier.

Reply:
My idea would be to add a simple check to `recycleByTransactionId` that `recycleProducer` is only invoked if `sucess` is true if false we close the producer. 

Afaict `recyleProducer` with success false is only called from `recycleByTransactionId` so the code path is only hit in a single scenario. 

This way we should be able to remove the parameter from `recyleProducer`.

Reply:
Okay I'll try to solve it in that way.

Reply:
Looks better this way. Thanks!